### PR TITLE
Authentication

### LINKED ISSUE #229: Introduce authentication and encryption for the network layer
Eventually we will transition the network layer to libp2p so we have a more structured way of designing network communication.

Up until then, we still want proper identification of peers (our current `TakerId` is transient with every connection), authentication of each other's messages and encryption of the entire communication.

To make the transition to libp2p later on easier, we first will:

- Use [`snow`](https://docs.rs/snow/0.8.0/snow/index.html) for the heavy lifting of encrypting and authenticating the connection
- Use the libp2p multi-addresses specification to configure the connection details to the other party

Using snow will require that we know the public key of the party ahead of time. Using libp2p's multiaddress provides several benefits:

- Specify connection details + public key in a single argument
- Allows usage of domains as well as IP addresses (https://github.com/comit-network/hermes/issues/292)

### PR DESCRIPTION
**Pull request by rishflab**
#229

### CODE REVIEW

File: start_all.sh [hunk 1]
```diff
@@ -4,4 +4,4 @@ export RUST_BACKTRACE=1
 
 # A simple command to spin up the complete package, ie. both daemons and frontends.
 # A single 'ctrl+c' stops all processes.
-(trap 'kill 0' SIGINT; cargo run --bin maker -- testnet & cargo run --bin taker -- testnet & APP=maker yarn --cwd=./frontend dev  & APP=taker yarn --cwd=./frontend dev)
+(trap 'kill 0' SIGINT; cargo run --bin maker -- testnet & cargo run --bin taker -- --maker-id 10d4ba2ac3f7a22da4009d813ff1bc3f404dfe2cc93a32bedf1512aa9951c95e testnet -- & APP=maker yarn --cwd=./frontend dev  & APP=taker yarn --cwd=./frontend dev)
```
Comment (klochowicz) [inline start_all.sh]: will this magic number (`maker_id`) ever change? Perhaps add a comment about it (either a reassuring one that it's set in stone, or how/when it needs to be updated)

File: daemon/src/noise.rs [hunk 2]
```diff
@@ -6,15 +6,19 @@ use tokio::net::TcpStream;
...
-// TODO: investigate what these params do and whether we can use get authentication for free using
-// them
-static NOISE_PARAMS: &str = "Noise_XX_25519_ChaChaPoly_BLAKE2s";
+static NOISE_PARAMS: &str = "Noise_IK_25519_ChaChaPoly_BLAKE2s";
 
-pub async fn initiator_handshake(connection: &mut TcpStream) -> Result<TransportState> {
+pub async fn initiator_handshake(
+    connection: &mut TcpStream,
+    local_priv_key: x25519_dalek::StaticSecret,
+    remote_pub_key: x25519_dalek::PublicKey,
```
Comment (klochowicz) [inline daemon/src/noise.rs]: nit: do you need to own these? ref would probably be enough.

File: daemon/src/noise.rs [hunk 3]
```diff
@@ -23,21 +27,22 @@ pub async fn initiator_handshake(connection: &mut TcpStream) -> Result<Transport
...
 
     tracing::debug!("Noise protocol initiator handshake is complete");
 
     Ok(noise)
 }
 
-pub async fn responder_handshake(connection: &mut TcpStream) -> Result<TransportState> {
+pub async fn responder_handshake(
+    connection: &mut TcpStream,
+    local_priv_key: x25519_dalek::StaticSecret,
```
Comment (klochowicz) [inline daemon/src/noise.rs]: nit: do you need to own this? (you could probably avoid a `clone()`)

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -76,10 +76,12 @@ jobs:
...
         shell: bash
         run: |
-          target/debug/maker --data-dir=/tmp/maker --generate-seed testnet &
+          cp daemon/util/testnet_seeds/maker_seed /tmp/maker/maker_seed
+          target/debug/maker --data-dir=/tmp/maker testnet &
           sleep 10s # Wait for maker to start
 
-          target/debug/taker --data-dir=/tmp/taker --generate-seed testnet &
+          cp daemon/util/testnet_seeds/taker_seed /tmp/taker/taker_seed:
+          target/debug/taker --data-dir=/tmp/taker --maker-id 10d4ba2ac3f7a22da4009d813ff1bc3f404dfe2cc93a32bedf1512aa9951c95e testnet &
```
Comment (klochowicz) [inline .github/workflows/ci.yml]: will we need to manually upgrade this key, increasing the maintenance burden?

Comment (rishflab) [reply .github/workflows/ci.yml]: if we change the seeds in the repo, yes

File: daemon/src/noise.rs (hunk 2 above)
Comment (rishflab) [reply daemon/src/noise.rs]: yep, ref is enough

File: daemon/src/taker.rs [hunk 5]
```diff
@@ -51,6 +55,12 @@ struct Opts {
     network: Network,
 }
 
+fn parse_x25519_pubkey(s: &str) -> Result<x25519_dalek::PublicKey> {
+    let mut bytes = [0u8; 32];
+    hex::decode_to_slice(s, &mut bytes)?;
+    Ok(x25519_dalek::PublicKey::from(bytes))
```
Comment (klochowicz) [inline daemon/src/taker.rs:61]: nit: `bytes.into()` ?

File: daemon/src/seed.rs [hunk 6]
```diff
@@ -72,6 +72,16 @@ impl Seed {
...
     }
+
+    pub fn derive_noise_static_secret(&self) -> x25519_dalek::StaticSecret {
+        let mut secret = [0u8; 32];
+
+        Hkdf::<Sha256>::new(None, &self.0)
+            .expand(b"NOISE_STATIC_SECRET", &mut secret)
+            .expect("okm array is of correct length");
+
+        x25519_dalek::StaticSecret::from(secret)
```
Comment (klochowicz) [inline daemon/src/seed.rs:83]: nit: `secret.into()`?

**Review by klochowicz — APPROVED** [review]
looks good!

**Pull request merged**